docs/2-features/17-oauth.md

```diff
@@ -0,0 +1,289 @@
+---
+title: OAuth
+description: "Learn how to implement OAuth to authenticate users with many different providers, such as GitHub, Google, Discord, and many others."
+keywords: "Experimental"
+---
+
+## Overview
+
+Tempest provides the ability to authenticate users with many OAuth providers, such as GitHub, Google, Discord, and many others, using the same interface.
+
+This implementation is built on top of the PHP league's [OAuth client](https://github.com/thephpleague/oauth2-client)—a reliable, battle-tested OAuth 2.0 client library.
+
+## Getting started
+
+To get started with OAuth, you will first need to create a configuration file for your desired OAuth provider.
+
+Tempest provides a [different configuration object for each provider](#available-providers). For instance, if you wish to authenticate users with GitHub, you may create a `github.config.php` file returning an instance of {b`Tempest\Auth\OAuth\Config\GitHubOAuthConfig`}:
+
+```php app/Auth/github.config.php
+return new GitHubOAuthConfig(
+    clientId: env('GITHUB_CLIENT_ID'),
+    clientSecret: env('GITHUB_CLIENT_SECRET'),
+    redirectTo: [GitHubOAuthController::class, 'callback'],
+    scopes: ['user:email'],
+);
+```
+
+In this example, the GitHub OAuth credentials are specified in the `.env`, so different credentials can be configured depending on the environment.
+
+Once your OAuth provider is configured, you may interact with it by using the {`Tempest\Auth\OAuth\OAuthClient`} interface. This is usually done through [dependency injection](../1-essentials/05-container.md#injecting-dependencies).
+
+## Implementing the OAuth flow
+
+To implement a complete OAuth flow for your application, you will need two routes.
+
+- The first one will redirect the user to the OAuth provider's authorization page,
+- The second one, which will be redirected to once the user authorizes your application, will fetch the user's information thanks to the code provided by the OAuth provider.
+
+The {b`Tempest\Auth\OAuth\OAuthClient`} interface has the necessary methods to handle both parts of the flow. The following is an example of a complete OAuth flow, including CSRF protection, creating or updating the user, and authenticating them against the application:
+
+```php app/Auth/DiscordOAuthController.php
+use Tempest\Auth\OAuth\OAuthClient;
+
+final readonly class DiscordOAuthController
```

Comment:
I haven't read further yet, but can we provide these controllers as discovery-skipped, installable ones? Something like `./tempest install oauth`, which lets you pick your providers, and then automatically makes the configs and controllers for you?

Reply:
Btw, this isn't a blocker for merging this, just an idea for a followup PR

Reply:
That sounds like a very cool idea! I can make another PR after this one

Reply:
Then let's merge when you're ready and follow up afterwards 👍 https://github.com/tempestphp/tempest-framework/issues/1582

If you want to, I can take care of the installer part as well